## Problem #345: Matrix Sum
[Link to Problem](https://projecteuler.net/problem=345)

### Problem Description

<p>We define the <dfn>Matrix Sum</dfn> of a matrix as the maximum possible sum of matrix elements such that none of the selected elements share the same row or column.</p>

<p>For example, the Matrix Sum of the matrix below equals 3315 ( = 863 + 383 + 343 + 959 + 767):</p>

<p style="text-align:center;font-family:'courier new';">
  7  53 183 439 <span style="color:#00ff00;font-family:'courier new';"><b>863</b></span><br>
497 <span style="color:#00ff00;font-family:'courier new';"><b>383</b></span> 563  79 973<br>
287  63 <span style="color:#00ff00;font-family:'courier new';"><b>343</b></span> 169 583<br>
627 343 773 <span style="color:#00ff00;font-family:'courier new';"><b>959</b></span> 943<br><span style="color:#00ff00;font-family:'courier new';"><b>767</b></span> 473 103 699 303<br></p>

<p>
Find the Matrix Sum of:<br></p>
<p style="text-align:center;font-family:'courier new';">
  7  53 183 439 863 497 383 563  79 973 287  63 343 169 583<br>
627 343 773 959 943 767 473 103 699 303 957 703 583 639 913<br>
447 283 463  29  23 487 463 993 119 883 327 493 423 159 743<br>
217 623   3 399 853 407 103 983  89 463 290 516 212 462 350<br>
960 376 682 962 300 780 486 502 912 800 250 346 172 812 350<br>
870 456 192 162 593 473 915  45 989 873 823 965 425 329 803<br>
973 965 905 919 133 673 665 235 509 613 673 815 165 992 326<br>
322 148 972 962 286 255 941 541 265 323 925 281 601  95 973<br>
445 721  11 525 473  65 511 164 138 672  18 428 154 448 848<br>
414 456 310 312 798 104 566 520 302 248 694 976 430 392 198<br>
184 829 373 181 631 101 969 613 840 740 778 458 284 760 390<br>
821 461 843 513  17 901 711 993 293 157 274  94 192 156 574<br>
 34 124   4 878 450 476 712 914 838 669 875 299 823 329 699<br>
815 559 813 459 522 788 168 586 966 232 308 833 251 631 107<br>
813 883 451 509 615  77 281 613 459 205 380 274 302  35 805<br></p>

### Approach

Basically we pick from each row an element and at the end no two such elements should be on the same column. Every valid matrix sum can be expressed as a permutation $P$, meaning we pick elements from cells $(1, P_1)$, $(2, P_2)$ and so on. If we bruteforce the permutations we gen time complexity $O(n!)$, which is too slow for a matrix of size $15 \times 15$.

A faster approach we can take is to use **dynamic programming**: $dp[i][mask]$ = the maximum sum we can get from the first $i$ rows choosing one element from each column included as a bit in $mask$. Thus we gen complexity $O(n \ 2^n)$

In [4]:
matrix = "  7  53 183 439 863 497 383 563  79 973 287  63 343 169 583\n" \
        "627 343 773 959 943 767 473 103 699 303 957 703 583 639 913\n" \
        "447 283 463  29  23 487 463 993 119 883 327 493 423 159 743\n" \
        "217 623   3 399 853 407 103 983  89 463 290 516 212 462 350\n" \
        "960 376 682 962 300 780 486 502 912 800 250 346 172 812 350\n" \
        "870 456 192 162 593 473 915  45 989 873 823 965 425 329 803\n" \
        "973 965 905 919 133 673 665 235 509 613 673 815 165 992 326\n" \
        "322 148 972 962 286 255 941 541 265 323 925 281 601  95 973\n" \
        "445 721  11 525 473  65 511 164 138 672  18 428 154 448 848\n" \
        "414 456 310 312 798 104 566 520 302 248 694 976 430 392 198\n" \
        "184 829 373 181 631 101 969 613 840 740 778 458 284 760 390\n" \
        "821 461 843 513  17 901 711 993 293 157 274  94 192 156 574\n" \
        " 34 124   4 878 450 476 712 914 838 669 875 299 823 329 699\n" \
        "815 559 813 459 522 788 168 586 966 232 308 833 251 631 107\n" \
        "813 883 451 509 615  77 281 613 459 205 380 274 302  35 805"

matrix = [list(map(int, line.split())) for line in matrix.splitlines()]

n = len(matrix)

dp = [[0 for _ in range(1 << n)] for _ in range(n)]

for j in range(n):
    dp[0][1 << j] = matrix[0][j]

for i in range(n - 1):
    for mask in range(1 << n):
        if dp[i][mask] > 0: # valid state
            for j in range(n): # try another bit
                if not (mask >> j) & 1: # that is not already included
                    dp[i + 1][mask | (1 << j)] = max(dp[i + 1][mask | (1 << j)], dp[i][mask] + matrix[i + 1][j])

print(dp[-1][-1])

13938


###### Result: **13938** | Execution time: ~0s

### Complexity analysis

Time complexity: $O(n^2 2^n)$

Space complexity: $O(n 2^n)$

##### Tags: #dynamic-programming, #bitmasks